### PR TITLE
Bloc pattern

### LINKED ISSUE #84: Bloc pattern documentation
There is a latex file that describes the bloc pattern. This needs to be cut down a bit and written in latex instead

### PR DESCRIPTION
**Pull request by Hylet**
This closes #84.
This is written by @aau-giraf/sw611f19 and edited by @aau-giraf/sw608f19.

### CODE REVIEW

**Comment by Tekrus** [comment]
I'd like @sipman to be one of the reviewers to ensure that he feels that this still represents the essentials of his original version

**Comment by Bjerkk** [comment]
@aau-giraf/sw611f19  you have been assigned this. Here's your checklist. I marked @sipman  as the reviewer, but you might change it as you like.
Please use the following checklist:
- [ ] Is there documented which BLoCs/Screens the current class is related to?

- [ ] Is the responsibility/functionality of the current class explained? Also regarding the classes it is related to?

- [ ] If any unusual decisions are introduced in the code are they reasoned about in the documentation?

Note: If you have any suggestions to what we can change, do not hesitate to contact us :D

**Comment by sipman** [comment]
@Bjerkk I will do the review, but your check list do not apply to this review :)

File: tutorials/bloc_pattern/BLoCPattern.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# BLoC Pattern
+When we moved the weekplanner application to Flutter, we chose to structure the code with the BLoC pattern.
+
+The name BLoC is used for two things, the BLoC pattern and classes called BLoC. To avoid confusion we always write BLoC pattern when we refer to the pattern. In all other cases we refer to the classes.
+
+The BLoC pattern is a pattern that aims to isolate all business logic from UI code in order to be able to re-use the business logic [6]. The business logic has to be platform independent, meaning usable on both the web-application and the mobile-application without any modifications [6].
```
Comment (sipman) [inline tutorials/bloc_pattern/BLoCPattern.md]: Not on the web-application and mobile as in only those. That is jsut the example set by google.

The pattern states that it use be useable on all platform.. So also desktop, embedded, what ever platform.

File: tutorials/bloc_pattern/BLoCPattern.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+## What is BLoC
+To understand the BLoC pattern one first needs to understand what a BLoC is. A BLoC is a simple class that only takes sinks as input and exposes streams as output [6]. All business logic should be inside a BLoC. The intuition is that BLoCs are the building blocks of the application [6]. Another key requirement of a BLoC is that all dependencies should be injected [6], so if an API is needed it should be injected in the constructor of the BLoC. This allows the web-platform to use a JSON-API and the mobile-platform to use a BINARY-API and other similar situations, while all platforms still uses the same BLoC without any modifications. An application thereby consist of many BLoC with different responsibilities. The BLoC can then be used by the UI components of all platforms.
```
Comment (sipman) [inline tutorials/bloc_pattern/BLoCPattern.md]:
```suggestion
To understand the BLoC pattern one first needs to understand what a BLoC is. A BLoC is a simple class that only takes sinks as input and exposes streams as output [6]. All business logic should be inside a BLoC. The intuition is that BLoCs are the building blocks of the application [6]. Another key requirement of a BLoC is that all dependencies should be injected [6], so if an API is needed it should be injected in the constructor of the BLoC. This allows the web-platform to use a JSON-API and the mobile-platform to use a BINARY-API and other similar situations, while all platforms still uses the same BLoC without any modifications. An application thereby consist of many BLoCs with different responsibilities. The BLoC can then be used by the UI components of all platforms.
```

File: tutorials/bloc_pattern/BLoCPattern.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+* UI components should be reactive, i.e. if data changes in one component then the change should automatically be reflected in all other components relying on the same data.
+* The pattern should allow for an intuitive file structure.
+
+Doing some empirical experiments in building with the different patterns and reading different developers experience with the different patterns we came to the conclusion that the BLoC pattern offered the most freedom while still isolating the business logic and UI logic. The BLoC pattern also allow for all the other goals.
+
+## BLoC Design Guidelines for the Giraf Project
+The Giraf Project have altered the rules of the BLoC pattern a bit. The reason for this is that in the Giraf Project we are not using the code on other platforms than the Flutter Framework. Therefore to make the use of BLoC a bit more intuitive and flexible for the developers, we have constructed the following design guideline:
+
+* Rather many small non-complex BLoCs than few large complex BLoCs
+* Start by creating a BLoC per UI Screen, then after implementing the functionality consider to refactor to shared BLoC
```
Comment (sipman) [inline tutorials/bloc_pattern/BLoCPattern.md]:
```suggestion
* Start by creating a BLoC per UI Screen, then after implementing the functionality consider to refactor to shared BLoCs
```

**Review by sipman — CHANGES_REQUESTED** [review]

File: tutorials/bloc_pattern/BLoCPattern.md (hunk 1 above)
Comment (Hylet) [reply tutorials/bloc_pattern/BLoCPattern.md]: Changed the examples to a more generic explanation

**Review by sipman — APPROVED** [review]

**Pull request merged**